File: packages/validator-go/validator/ternary.go
package validator

import (
	"regexp"
	"strconv"
	"strings"
)

var (
	ternaryRe     = regexp.MustCompile(`\?[^:]*:`)
	conditionOpRe = regexp.MustCompile(`(==|!=|>=|<=|>|<|\s+in\s+|\s+not\s+in\s+)`)
)

// IsTernaryExpression reports whether a string contains a top-level
// "cond ? a : b" ternary form.
func IsTernaryExpression(value string) bool {
	return ternaryRe.MatchString(value)
}

// IsConditionExpression reports whether a string looks like a condition
// expression (leading dot path reference or comparison operators).
// Ternary expressions are excluded; evaluate those with EvaluateTernaryString.
func IsConditionExpression(value string) bool {
	if IsTernaryExpression(value) {
		return false
	}
	return strings.HasPrefix(value, ".") || conditionOpRe.MatchString(value)
}

// EvaluateTernaryString evaluates a "cond ? a : b" expression textually.
// The condition is evaluated with the condition parser; the chosen branch is
// returned as a literal value (number/bool/null/string) without re-parsing,
// so branches may contain arbitrary text such as regex patterns.
func (cp *ConditionParser) EvaluateTernaryString(expression string, formData map[string]interface{}, currentPath []string) interface{} {
	expression = strings.TrimSpace(expression)

	questionPos := findTernaryOperator(expression, '?', 0)
	if questionPos == -1 {
		// Not a ternary: evaluate as a regular condition
		result, err := cp.Evaluate(expression, formData, currentPath)
		if err != nil {
			return nil
		}
		return result
	}

	colonPos := findTernaryOperator(expression, ':', questionPos+1)
	if colonPos == -1 {
		return nil // malformed ternary
	}

	condition := strings.TrimSpace(expression[:questionPos])
	trueValue := strings.TrimSpace(expression[questionPos+1 : colonPos])
	falseValue := strings.TrimSpace(expression[colonPos+1:])

	conditionResult, err := cp.Evaluate(condition, formData, currentPath)
	if err != nil {
		conditionResult = false
	}

	resultExpr := falseValue
	if conditionResult {
		resultExpr = trueValue
	}

	// Nested ternary
	if IsTernaryExpression(resultExpr) {
		return cp.EvaluateTernaryString(resultExpr, formData, currentPath)
	}

	return parseTernaryBranchValue(resultExpr)
}

// findTernaryOperator finds the position of a top-level '?' or ':',
// respecting quotes, parentheses/brackets, and nested ternaries.
func findTernaryOperator(expression string, operator byte, startPos int) int {
	depth := 0
	inQuote := false
	var quoteChar byte
	ternaryDepth := 0

	for i := startPos; i < len(expression); i++ {
		ch := expression[i]

		// Handle quotes
		if (ch == '"' || ch == '\'') && !inQuote {
			inQuote = true
			quoteChar = ch
		} else if inQuote && ch == quoteChar {
			inQuote = false
			quoteChar = 0
		}

		if inQuote {
			continue
		}

		switch ch {
		case '(', '[':
			depth++
		case ')', ']':
			depth--
		case '?':
			if depth == 0 {
				if operator == '?' {
					return i
				}
				ternaryDepth++
			}
		case ':':
			if depth == 0 && operator == ':' {
				if ternaryDepth == 0 {
					return i
				}
				ternaryDepth--
			}
		}
	}

	return -1
}

// parseTernaryBranchValue parses a ternary branch as a literal value.
func parseTernaryBranchValue(value string) interface{} {
	value = strings.TrimSpace(value)

	// Bracket-enclosed list (e.g., [US, CA, UK])
	if len(value) >= 2 && strings.HasPrefix(value, "[") && strings.HasSuffix(value, "]") {
		inner := strings.TrimSpace(value[1 : len(value)-1])
		parts := splitValueList(inner)
		result := make([]interface{}, 0, len(parts))
		for _, p := range parts {
			result = append(result, parseTernaryBranchValue(p))
		}
		return result
	}

	// Quoted strings
	if len(value) >= 2 {
		first := value[0]
		last := value[len(value)-1]
		if (first == '"' && last == '"') || (first == '\'' && last == '\'') {
			return value[1 : len(value)-1]
		}
	}

	// Comma-separated list (without brackets)
	if strings.Contains(value, ",") && !strings.Contains(value, "[") {
		parts := splitValueList(value)
		result := make([]interface{}, 0, len(parts))
		for _, p := range parts {
			result = append(result, parseTernaryBranchValue(p))
		}
		return result
	}

	switch value {
	case "true":
		return true
	case "false":
		return false
	case "null":
		return nil
	}

	if num, err := strconv.ParseFloat(value, 64); err == nil {
		return num
	}

	return value
}

// splitValueList splits a comma-separated value list respecting quotes.
func splitValueList(value string) []string {
	var parts []string
	var sb strings.Builder
	inQuote := false
	var quoteChar byte

	for i := 0; i < len(value); i++ {
		ch := value[i]
		if (ch == '"' || ch == '\'') && !inQuote {
			inQuote = true
			quoteChar = ch
		} else if inQuote && ch == quoteChar {
			inQuote = false
			quoteChar = 0
		}

		if ch == ',' && !inQuote {
			parts = append(parts, strings.TrimSpace(sb.String()))
			sb.Reset()
			continue
		}
		sb.WriteByte(ch)
	}
	parts = append(parts, strings.TrimSpace(sb.String()))
	return parts
}

File: packages/validator-rust/src/legacy/ternary.rs
//! Textual ternary evaluation and condition/ternary detection.
//! Ports validator-go/validator/ternary.go.

use crate::legacy::condition_parser::ConditionParser;
use regex::Regex;
use serde_json::Value;
use std::sync::OnceLock;

fn ternary_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"\?[^:]*:").unwrap())
}

fn condition_op_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(==|!=|>=|<=|>|<|\s+in\s+|\s+not\s+in\s+)").unwrap())
}

/// is_ternary_expression reports whether a string contains a top-level
/// "cond ? a : b" form (matches Go regexp `\?[^:]*:`).
pub fn is_ternary_expression(value: &str) -> bool {
    ternary_re().is_match(value)
}

/// is_condition_expression reports whether a string looks like a condition:
/// a leading dot path reference or a comparison operator. Ternaries excluded.
pub fn is_condition_expression(value: &str) -> bool {
    if is_ternary_expression(value) {
        return false;
    }
    value.starts_with('.') || condition_op_re().is_match(value)
}

/// evaluate_ternary_string evaluates a "cond ? a : b" expression textually.
/// The condition is evaluated with the parser; the chosen branch is returned
/// as a literal value without re-parsing, so branches may contain arbitrary
/// text (e.g. regex patterns). Returns None on malformed input.
pub fn evaluate_ternary_string(
    cp: &mut ConditionParser,
    expression: &str,
    form_data: &Value,
    current_path: &[String],
) -> Option<Value> {
    let expression = expression.trim();

    let question_pos = find_ternary_operator(expression, b'?', 0);
    if question_pos.is_none() {
        // Not a ternary: evaluate as a regular condition.
        return match cp.evaluate(expression, form_data, current_path) {
            Ok(b) => Some(Value::Bool(b)),
            Err(_) => None,
        };
    }
    let question_pos = question_pos.unwrap();

    let colon_pos = find_ternary_operator(expression, b':', question_pos + 1)?;

    let condition = expression[..question_pos].trim();
    let true_value = expression[question_pos + 1..colon_pos].trim();
    let false_value = expression[colon_pos + 1..].trim();

    let condition_result = cp
        .evaluate(condition, form_data, current_path)
        .unwrap_or(false);

    let result_expr = if condition_result {
        true_value
    } else {
        false_value
    };

    if is_ternary_expression(result_expr) {
        return evaluate_ternary_string(cp, result_expr, form_data, current_path);
    }

    Some(parse_ternary_branch_value(result_expr))
}

/// find_ternary_operator finds a top-level '?' or ':' respecting quotes,
/// parentheses/brackets, and nested ternaries. Indices are byte positions.
fn find_ternary_operator(expression: &str, operator: u8, start_pos: usize) -> Option<usize> {
    let bytes = expression.as_bytes();
    let mut depth = 0i32;
    let mut in_quote = false;
    let mut quote_char = 0u8;
    let mut ternary_depth = 0i32;

    let mut i = start_pos;
    while i < bytes.len() {
        let ch = bytes[i];

        if (ch == b'"' || ch == b'\'') && !in_quote {
            in_quote = true;
            quote_char = ch;
        } else if in_quote && ch == quote_char {
            in_quote = false;
            quote_char = 0;
        }

        if in_quote {
            i += 1;
            continue;
        }

        match ch {
            b'(' | b'[' => depth += 1,
            b')' | b']' => depth -= 1,
            b'?' => {
                if depth == 0 {
                    if operator == b'?' {
                        return Some(i);
                    }
                    ternary_depth += 1;
                }
            }
            b':' => {
                if depth == 0 && operator == b':' {
                    if ternary_depth == 0 {
                        return Some(i);
                    }
                    ternary_depth -= 1;
                }
            }
            _ => {}
        }
        i += 1;
    }
    None
}

/// parse_ternary_branch_value parses a ternary branch as a literal value:
/// bracket list, quoted string, comma list, true/false/null, number, or
/// bare string.
pub fn parse_ternary_branch_value(value: &str) -> Value {
    let value = value.trim();

    // Bracket-enclosed list (e.g. [US, CA, UK]).
    if value.len() >= 2 && value.starts_with('[') && value.ends_with(']') {
        let inner = value[1..value.len() - 1].trim();
        let parts = split_value_list(inner);
        let result: Vec<Value> = parts
            .iter()
            .map(|p| parse_ternary_branch_value(p))
            .collect();
        return Value::Array(result);
    }

    // Quoted strings.
    if value.len() >= 2 {
        let bytes = value.as_bytes();
        let first = bytes[0];
        let last = bytes[value.len() - 1];
        if (first == b'"' && last == b'"') || (first == b'\'' && last == b'\'') {
            return Value::String(value[1..value.len() - 1].to_string());
        }
    }

    // Comma-separated list without brackets.
    if value.contains(',') && !value.contains('[') {
        let parts = split_value_list(value);
        let result: Vec<Value> = parts
            .iter()
            .map(|p| parse_ternary_branch_value(p))
            .collect();
        return Value::Array(result);
    }

    match value {
        "true" => return Value::Bool(true),
        "false" => return Value::Bool(false),
        "null" => return Value::Null,
        _ => {}
    }

    if let Ok(num) = value.parse::<f64>() {
        return serde_json::json!(num);
    }

    Value::String(value.to_string())
}

/// split_value_list splits a comma-separated list respecting quotes.
fn split_value_list(value: &str) -> Vec<String> {
    let bytes = value.as_bytes();
    let mut parts = Vec::new();
    let mut sb = String::new();
    let mut in_quote = false;
    let mut quote_char = 0u8;

    let mut i = 0;
    while i < bytes.len() {
        let ch = bytes[i];
        if (ch == b'"' || ch == b'\'') && !in_quote {
            in_quote = true;
            quote_char = ch;
        } else if in_quote && ch == quote_char {
            in_quote = false;
            quote_char = 0;
        }

        if ch == b',' && !in_quote {
            parts.push(sb.trim().to_string());
            sb.clear();
            i += 1;
            continue;
        }
        sb.push(ch as char);
        i += 1;
    }
    parts.push(sb.trim().to_string());
    parts
}
